rfcs/0000-top-level-assumptions.md

```diff
@@ -0,0 +1,315 @@
+
+- We could decide to add a simple top-level `loose` option, like we did in Babel 5: this solves the problem of multiple plugins sharing the same assumption. However, it would be even more obscure than the current `loose` options inside the plugins (which at least give some context). Also, having an all-or-nothing toggle means that often you would have to disable _every_ optimization just because you rely on a single edge case.
+
+# Adoption strategy
+
+We can introduce these options without introducing breaking changes, because it's opt-in. There are three possible compatibility strategies between the existing `loose` options and `assumptions` that we can choose:
+
+1. Allow both, `assumptions` has higher precedence because it's the new recommended option:
+    ```javascript
+    export default function transformClassProperties(api, options) {
+      const { setPublicClassFields = options.loose } = api.assumptions;
+    ```
+
+1. Allow both, `loose` has higher precedence because it "nearer" to the plugin:
+    ```javascript
+    export default function transformClassProperties(api, options) {
+      const setPublicClassFields = options.loose ?? api.assumptions.setPublicClassFields;
+    ```
+
+1. Disallow using both at the same time. This will make the migration faster but a little harder:
+    ```javascript
+    export default function transformClassProperties(api, options) {
+      let { setPublicClassFields } = options.loose;
+      if (options.loose != null) {
+        if (setPublicClassFields != null) throw new Error();
+        setPublicClassFields = options.loose;
+      }
+    ```
+
+I prefer the first alternative, since it makes it possible for our users to gradually migrate to the new options and gradually makes the `loose` options noops in users' configs while migrating.
+
+# How we teach this
+
+This RFC proposed the addition of a new set of options in a way that every option can affect multiple plugins, and a plugin can be affected by multiple options.
+
+Documenting the different `loose` and `loose`-like options wasn't too hard: every plugin had a bunch of those options, and we could document them in the plugin's page.
+
+A possible structure for the documentation could be a two-entry table, with all the plugins in the rows and the assumptions in the columns, marking which assumption affects which plugins. However, this would lead to an enormous highly sparse table, impossible to fit in a normal web page.
+
+I think that the best way of documenting these options is to put all of them in a single long list, divided in two sections: assumptions that affect standard features, and assumptions that only affect proposals. For each assumption we should write:
+- The option name
+- It's description
+- An example (which could be interactive, using CodeSandbox)
+- The affected plugins
```

Comment:
Maybe worthy to offer a green engines list for each assumptions. Like if you are _not_ transpiling down to these browsers, this assumption is a no-op for you. i.e. People should not worry about what is `setSpreadProperties` if they are not traspiling object spread anyway.

Reply:
We could even automatically generate it by computing the union of all the affected plugins' targets